contracts/AMM/LiquidityPool/ThalesAMMLiquidityPool.sol

```diff
@@ -0,0 +1,716 @@
+
+            (IPosition up, IPosition down) = IPositionalMarket(market).getOptions();
+            IPosition target = position == IThalesAMM.Position.Up ? up : down;
+
+            ThalesAMMLiquidityPoolRound(liquidityPoolRound).moveOptions(
+                IERC20Upgradeable(address(target)),
+                optionsAmount,
+                address(thalesAMM)
+            );
+        }
+    }
+
+    /// @notice get options that are in the LP into the AMM for the buy tx
+    /// @param market to get options for
+    /// @param optionsAmount to get options for
+    /// @param position to get options for
+    function getOptionsForBuyByAddress(
+        address market,
+        uint optionsAmount,
+        address position
+    ) external nonReentrant whenNotPaused onlyAMM {
+        if (optionsAmount > 0) {
+            require(started, "Pool has not started");
+
+            uint marketRound = getMarketRound(market);
+            address liquidityPoolRound = _getOrCreateRoundPool(marketRound);
+
+            ThalesAMMLiquidityPoolRound(liquidityPoolRound).moveOptions(
+                IERC20Upgradeable(position),
+                optionsAmount,
+                address(thalesAMM)
+            );
+        }
+    }
+
+    /// @notice request withdrawal from the LP
+    /// @param market to check
+    /// @return roundPool the pool for the passed market
+    function getOrCreateMarketPool(address market) external onlyAMM nonReentrant whenNotPaused returns (address roundPool) {
+        uint marketRound = getMarketRound(market);
+        roundPool = _getOrCreateRoundPool(marketRound);
+    }
+
+    /// @notice request withdrawal from the LP
+    function withdrawalRequest() external nonReentrant whenNotPaused {
+        require(started, "Pool has not started");
+        require(!withdrawalRequested[msg.sender], "Withdrawal already requested");
+        require(balancesPerRound[round][msg.sender] > 0, "Nothing to withdraw");
+        require(balancesPerRound[round + 1][msg.sender] == 0, "Can't withdraw as you already deposited for next round");
+
+        if (!whitelistedDeposits[msg.sender]) {
+            require(
+                balancesPerRound[round][msg.sender] <
+                    _transformCollateral(((stakingThales.stakedBalanceOf(msg.sender) * stakedThalesMultiplier) / ONE)),
+                "Not enough staked THALES"
+            );
+        }
+
+        if (totalDeposited > balancesPerRound[round][msg.sender]) {
+            totalDeposited -= balancesPerRound[round][msg.sender];
+        } else {
+            totalDeposited = 0;
+        }
+
+        usersCurrentlyInPool = usersCurrentlyInPool - 1;
+        withdrawalRequested[msg.sender] = true;
+        emit WithdrawalRequested(msg.sender);
+    }
+
+    /// @notice Close current round and begin next round,
+    /// excercise options of trading markets and calculate profit and loss
+    function closeRound() external nonReentrant whenNotPaused {
```

Comment:
Please take the latest code from SportAMMLiquidityPool that includes batching for round closing